### ISSUE TITLE
A change to Z-Probing?

### ISSUE THREAD

**Issue by Mendatenrognabs** [luc-github/Repetier-Firmware-4-Davinci#288]
Z-probing doesn't work as I expected after upgrading firmware and returns: 'Error:Z-probe failed' or 'Error:z-probe triggered before starting probing'.

Printer: DaVinci 2.0A (with J37 jumper)
Initial Firmware: Repetier 0.92 (luc 30 Oct 2015)
Current Firmware: Repetier 0.92.10 (luc 17 Nov 2017)

After playing with both manual and automatic bed leveling/z-probing I realized that getting a trued up bed could be much easier.  If the three probing points were located directly above the three bed leveling screws, then any error in a point's measurement can be corrected by only changing that point's leveling screw.  This can be done without effecting the measurements at the other two points.

Of course, the built in Z-probe can't be used to measure at these three critical points.  To overcome this limitation of the printer hardware, I made a simple modification.  Applying aluminum foil tape to the entire bed allows the nozzle to complete the electrical circuit and measure at any point on the bed.  While applying the tape, take extra care that it is perfectly smooth with no bubbles or wrinkles and confirm that there is continuity between the tape and the blue wires attached to the bed.  Once the bed is trued up, using some blue masking tape on top of the foil gives good print adhesion.

This approach works pretty good with the 0.92 firmware that I installed shortly after purchasing the printer a few years back.  Using Repetier host I change the EEPROM values that set the X and Y positions so that the three points are above the leveling screws.  To probe, I send G28 and then G29.  The printer moves to the first point/level screw, moves the bed down 5 mm, then slowly lifts the bed up until electrical contact and the measurement are made.  It then moves to the next point and repeats the process.  Looking at the results I can quickly determine which screw(s) need to change as well as which direction to change them and by about how much.  After trueing up the bed, I can send G28 followed by G32 to calculate the bed rotation and it works in much the same way.

Realizing that my firmware was two years out of date I decided to upgrade it.  After the upgrade I attempted to use the same process for bed leveling.  Foil tape in place, G28, G29.  The results are less than glamourous.  After G28, the printer moves to the appropriate x,y position, drops down and returns the error "Z-probe failed".  It does not appear that the bed even moves upward to attempt the measurement.  Here is a sample log:
 

> 11:21:51.602 : N18 G28*42
11:21:51.629 : SelectExtruder:0
11:21:52.815 : N19 M105*31
11:21:55.903 : N20 M105*21
11:21:57.660 : X:0.00 Y:0.00 Z:0.000 E:0.0000
11:21:59.003 : N21 M105*20
11:22:02.076 : N22 M105*23
11:22:05.181 : N23 M105*22
11:22:06.909 : N24 G29*36
11:22:08.284 : N25 M105*16
11:22:11.113 : Error:Z-probe failed
11:22:11.169 : X:118.50 Y:40.00 Z:-400.000 E:0.0000
11:22:11.372 : N26 M105*19
11:22:13.734 : fatal:G29 leveling failed! Printer stopped and heaters disabled due to this error. Fix error and restart with M999.
11:22:13.753 : SelectExtruder:0
11:22:26.242 : X:0.00 Y:0.00 Z:0.000 E:0.0000
11:22:26.242 : fatal:G29 leveling failed! Printer stopped and heaters disabled due to this error. Fix error and restart with M999.
11:22:26.242 : fatal:G29 leveling failed! Printer stopped and heaters disabled due to this error. Fix error and restart with M999.
11:22:27.459 : N27 M999*31
11:22:27.459 : N1 M110*34
11:22:27.461 : DebugLevel:6
11:22:27.463 : info:Continue from fatal state
11:22:28.466 : N1 M110*34

The day that I made the upgrade, I was successfully z-probing with 0.92, took about an hour to complete the upgrade (it had been years since I last did it), and then immediately tried setting the appropriate points and running G29 without success.

While diagnosing the issue, I have determined that if I home, then move to the correct x,y points and down 10 mm I can use G30 to successfully probe a single point.  I can then follow with G28 and then G29 and receive the 'Error:z-probe triggered before starting probing' message.  Lowering the bed at that point by turning the bedlevel screw ~23° then causes G30 to return 'Error:Z-probe failed'.

While I am no programmer, I did compare the two versions of the code enough to know that there has been a major overhaul of the relevant sections.  Any ideas on what might be happening and how to fix it are welcome.

Finally, thank you for all the hard work on the firmware, it is a huge improvement over stock.

**Comment by luc-github** [comment]
the latest version of FW is based on 0.92.10 from original repetier which is supposed to be the correct one for autoleveling as it was fully rewritten - I guess you noticed it 

It is a long time I did not use autoleveling but it was working - first your bed must be mannualy leveled using the screws, then you can use the autoleveling to adjust the final leveling.
 that said the corner plates and the probe must be very clean to allow good contact and contact position must be on corner plate - if not it will fail

**Comment by Mendatenrognabs** [comment]
Thanks for the response.  Because I have added the foil tape to the bed and changed Z-probe X1...Y3 I am not using the three corner plates and the pin on the right extruder for z-probing.  Instead I am using the right extruder's nozzle and the foil tape covered bed.  This allows for z-probing at any location on the bed.

I know that I have good electrical contact because:
1. under some conditions I can get the z-probe (G30) to work at the appropriate location and immediately follow up with the G29 command and have it fail.
2. The G29 command was working under the 30 Oct 2015 0.92 firmware and then, an hour or so later and without changing anything except the firmware, the G29 command did not work.

Some additional observations after playing with it this afternoon: 
* Using the lcd menu the Maintenance->Zmin evaluation and Maintenance-> Auto level bed commands both work.
* When using Repetier host, I can home and then send G1 Z5 (or greater) followed by G29 and it works.

My understanding is that to properly use G29, one must first home the printer (G28).

I think what is happening is that with the 0.92.10 firmware the G29 command checks to see if the current Z position is greater than 'max zprobe - bed dist.' (from EEPROM=5 mm).  If it is greater, then the bed moves up to 'max zprobe - bed dist.', if it is less or equal then it moves to the next step.  The next step is moving to 'z-probe X1', 'z-probe Y1' (from EEPROM=118.5, 40).  I believe that the z-probe pin goes from low to high during this move if I have sent G28, G29 because the nozzle makes contact with the bed during the move.

With the 0.92 firmware the behavior was that the G29 command moved the bed down to 'max zprobe - bed dist.' if it was less than the value and moved to the next step if it was greater than or equal to the 'max zprobe - bed dist.' value.

Short version:  The G29 and G32 commands used to, starting from home, move the bed down, then move to the correct probing position then probe.  Now they, starting from home, don't change the bed height, and try to move to the correct position making contact with the foil covered bed and causing the command to fail without ever probing.

Am I understanding the documentation correctly that G28 must be entered before G29 and that no moves can be made between?

**Comment by luc-github** [comment]
you need to take in consideration the corner plate and foil are not in same heigh so this will make the calculation false and introduce too much distance in calculation - nozzle and probe are not in same heigh neither which are more issue in calculation accuracy and may generate the error

**Comment by Mendatenrognabs** [comment]
To take into consideration the difference in height between corner plate and foil I set Z_PROBE_HEIGHT 0 (I should have mentioned that earlier, sorry!)

After giving the documentation at https://www.repetier.com/documentation/repetier-firmware/z-probing/ a careful read, I believe that what I am seeing is exactly the same as what they claim it should do.  Its pretty clear that Repetier is coded for endstops at Z-max not Z-min.  Not moving the bed away from the nozzle when configured with Z-min endstops seems like a flaw, but is the intended design.

Working around it isn't too hard.  I changed configuration.h from: #define Z_PROBE_START_SCRIPT ""
to: #define Z_PROBE_START_SCRIPT "G0 Z5"

Initial testing suggest that this works - I'll post again if I run into any unexpected problems.

Thank you for the help!

**Comment by luc-github** [comment]
I guess issue is solved and issue can be closed
